hazelcast/src/main/java/com/hazelcast/sql/impl/expression/ParameterExpression.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Dynamic parameter expression.
+ */
+public final class ParameterExpression<T> implements Expression<T>, IdentifiedDataSerializable {
+
+    private int index;
+    private QueryDataType type;
+
+    public ParameterExpression() {
+        // No-op.
+    }
+
+    private ParameterExpression(int index, QueryDataType type) {
+        this.index = index;
+        this.type = type;
+    }
+
+    public static ParameterExpression<?> create(int index, QueryDataType type) {
+        return new ParameterExpression<>(index, type);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_PARAMETER;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        return (T) context.getArgument(index);
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return type;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeInt(index);
+        out.writeObject(type);
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        index = in.readInt();
+        type = in.readObject();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(index);
```

Comment:
Type is not used for hash code calculation but is used in `equals`.

Reply:
fixed